src/node_api_helpers.h

```diff
@@ -38,6 +45,289 @@
+    napi_env env = napi_get_current_env();
+    napi_propertyname key = reinterpret_cast<napi_propertyname>(name);
+    napi_set_property(env, obj, key, val);
+  }
+  inline void Set(napi_value arr, uint32_t i, napi_value val) {
+    napi_env env = napi_get_current_env();
+    napi_set_element(env, arr, i, val);
+  }
+
+  // Type check helpers
+  inline bool IsEmpty(napi_value v) {
+    napi_env env = napi_get_current_env();
+    return napi_is_empty(env, v);
+  }
+  inline bool IsNumber(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_number;
+  }
+  inline bool IsInt32(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_is_int32(env, val);
+  }
+  inline bool IsString(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_string;
+  }
+  inline bool IsFunction(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_function;
+  }
+  inline bool IsObject(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_object;
+  }
+  inline bool IsArray(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_array;
+  }
+  inline bool IsBoolean(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_boolean;
+  }
+  inline bool IsUndefined(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_undefined;
+  }
+  inline bool IsSymbol(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_symbol;
+  }
+  inline bool IsRegExp(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_regexp;
+  }
+  inline bool IsDate(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_date;
+  }
+  inline bool IsNull(napi_value val) {
+    napi_env env = napi_get_current_env();
+    return napi_get_type_of_value(env, val) == napi_null;
+  }
+
+  inline bool HasInstance(napi_value tpl, napi_value obj) {
+    napi_env env = napi_get_current_env();
+    return napi_instanceof(env, obj, tpl);
+  }
+
+  inline napi_value CopyBuffer(const char* buf, uint32_t size) {
+    napi_env env = napi_get_current_env();
+    return napi_buffer_copy(env, buf, size);
+  }
+
+  inline napi_value StringConcat(napi_value str1, napi_value str2) {
+    napi_env env = napi_get_current_env();
+    return napi_string_concat(env, str1, str2);
+  }
+
+  inline napi_value MakeCallback(
+      napi_value target,
+      napi_value func,
+      int argc,
+      napi_value* argv
+  ) {
+    napi_env env = napi_get_current_env();
```

Comment:
All these helpers remind me of something I haven't understood yet... if it is OK to always use the "current" environment (isolate), then why do NAPI functions all have the explicit environment parameter? Nan always just uses the current one, right? In what scenario might a caller of a NAPI function specify an environment that is different from the current one?